Ex2/Ex2.java

```diff
@@ -0,0 +1,209 @@
+import java.util.HashSet;
+import java.util.Random;
+
+public class Ex2 {
+
+	public static final boolean LEFT = true;
+	public static final boolean RIGHT = false;
+	
+	public static void main(String[] args) {
+		System.out.println("This program actually prints the tree so results can be seen.");
+		System.out.println("Uncomment one of these two lines in \"main\" to check a tree:");
+		System.out.println("//checkWithRandomTree();");
+		System.out.println("//checkWithAssignmentTree();");
+		//checkWithRandomTree();
+		//checkWithAssignmentTree();
+	}
+	
+	/**
+	 * This method prints the ancestors of a given node in the Binary Tree
+	 * @param tree - The Binary Tree Data Structure to work on
+	 * @param targetNode - the node which we would like to print its ancestors
+	 */
+	public static void printAncestors(BinaryTree<?> tree, TreeNode<?> targetNode) {
+		if (tree == null || tree.getRoot() == null) {
+			System.out.println("Tree is not initialized yet.");
+			return;
+		}
+		
+		if (targetNode == null) {
+			System.out.println("Node is not initialized yet.");
+			return;
+		}
+		
+		if (!tree.exists(targetNode)) {
+			System.out.println("No such node in the tree.");
+			return;
+		}
+		
+		System.out.println("Ancestors list of key " + targetNode + ":");
+		
+		if (tree.getRoot().getData() == targetNode.getData()) {
+			System.out.println("No ancestors! you're at the root!");
+			return;
+		}
+		
+		printAncestors(tree.getRoot(), targetNode);
+		System.out.println("|");
+		System.out.println();
+	}
+	
+	/**
+	 * A recursive function that outputs the ancestors of a given node.
+	 * The function iterates on the "next" node until it reaches the desired value
+	 * and then retrace its ancestors in the correct order
+	 * @param nextNode - a TreeNode to iterate on
+	 * @param targetNode - the desired node to find
+	 */
+	private static void printAncestors(TreeNode<?> nextNode, TreeNode<?> targetNode) {
+		if (nextNode.getData() == targetNode.getData()) {
+			System.out.print("Key - ");
+		}
+		
+		boolean hasRight = (nextNode.getRight() != null);
+		boolean hasLeft = (nextNode.getLeft() != null);
+		
+		if (hasRight && nextNode.getRight().exists(targetNode)) {
```

Comment:
Think how many times each node will be visited with this algorithm. I think it might be n^2. Can you do better?